### PR TITLE
Add MySQL capability

### LINKED ISSUE #76: Implement mysql/mariadb capability

### PR DESCRIPTION
**Pull request by realsdx**
FIxes #76

### CODE REVIEW

**Comment by kajojify** [comment]
Great! Well done! :)

**Comment by kajojify** [comment]
Please, get rid of redundant parentheses around conditions in if-statements.

**Comment by realsdx** [comment]
> Please, get rid of redundant parentheses around conditions in if-statements.

Sorry! I have a bad habit of using parentheses

**Comment by johnnykv** [comment]
Great job, but this really need a test case. Take a look at the test case for postgresql.

**Comment by kajojify** [comment]
I had a few suggestions :) 
will write a bit later

**Comment by realsdx** [comment]
Okay, I will start working on the test case

**Comment by kajojify** [comment]
We definetely need some links to MySQL docs. So, we will be able to understand, what goes on, if we want. Put them before MySQL declaration. Just example:
```
# Authentication was implemented according to <link1> and <link2>
class MySQL(HanlderBase):
    ...
```

**Comment by johnnykv** [comment]
I agree with @kajojify, some code comments would be good.

**Comment by realsdx** [comment]
Okay, in a draft branch i already have comments for almost everything, i will update it soon.

**Comment by kajojify** [comment]
Why do you think we need to use hexadecimal values in many places?

**Comment by realsdx** [comment]
In docs most of the values are in hex . Also, i was working with bytes , so i thought it might be helpful

**Comment by johnnykv** [comment]
Did you test with hydra?

When testing with` hydra -l someUser -p somePass mysql://127.0.0.1`

I get the following exception

```
2019-03-26 18:49:15,648 (asyncio) Task exception was never retrieved
future: <Task finished coro=<HandlerBase.handle_session() done, defined at /home/johnny/repos/heralding/heralding/capabilities/handlerbase.py:74> exception=ValueError('subsection not found',)>
Traceback (most recent call last):
  File "/home/johnny/repos/heralding/heralding/capabilities/handlerbase.py", line 87, in handle_session
    timeout=self.timeout, loop=self.loop)
  File "/usr/lib/python3.6/asyncio/tasks.py", line 358, in wait_for
    return fut.result()
  File "/home/johnny/repos/heralding/heralding/capabilities/mysql.py", line 126, in execute_capability
    await self._handle_session(reader, writer, session)
  File "/home/johnny/repos/heralding/heralding/capabilities/mysql.py", line 153, in _handle_session
    username_end_pos = data.index(b'\x00', 0x24, max_size-0x24)
ValueError: subsection not found
```

Note that the credentials are logged alright...

**Comment by johnnykv** [comment]
My guess is that this occurs because hydra done not honor the protocol on first connect. The exception can be reproduced by `telnet 127.0.0.1 3306`...

**Comment by realsdx** [comment]
No, i didn't use hydra. I used simple bruteforce script (built using pymysql) to test. 
Exception handling required here.

**Comment by realsdx** [comment]
I tested it, i think it has something to do with concurrent session in asyncio. I will research more about it.

**Comment by kajojify** [comment]
If you build some new capability, please try to make it workable with all used brute forcers - at least hydra, medusa and ncrack. John the ripper additionaly

File: heralding/capabilities/mysql.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        # XX+1 ??      Password
+        # YY   ??      Schema (NUL-terminated)
+        # ZZ   ??      Client authentication plugin (NUL-terminated)
+        # RR   ??      Client connection attibutes
+
+        address = writer.get_extra_info('peername')[0]
+        writer.write(self.server_greeting())
+        data = await reader.read(2048)
+
+        caps = int.from_bytes(data[0x04:0x08], byteorder='little')
```
Comment (johnnykv) [inline heralding/capabilities/mysql.py]: You need to keep track of the length of data. Here we are using indexing (0x04), but are we sure that data even holds that many elements?

File: heralding/capabilities/mysql.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        max_size = int.from_bytes(data[0x08:0x0C], byteorder='little')
```
Comment (johnnykv) [inline heralding/capabilities/mysql.py]: The same thing here, and many other places :-) .

**Comment by realsdx** [comment]
I updated the PR. Please review it.

**Pull request merged**